// xgplayer-hls-live/lib/mobile/config.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _xgplayerTransmuxerBufferTrack = require('xgplayer-transmuxer-buffer-track');

var _xgplayerTransmuxerBufferTrack2 = _interopRequireDefault(_xgplayerTransmuxerBufferTrack);

var _xgplayerTransmuxerBufferXgbuffer = require('xgplayer-transmuxer-buffer-xgbuffer');

var _xgplayerTransmuxerBufferXgbuffer2 = _interopRequireDefault(_xgplayerTransmuxerBufferXgbuffer);

var _xgplayerTransmuxerLoaderFetch = require('xgplayer-transmuxer-loader-fetch');

var _xgplayerTransmuxerLoaderFetch2 = _interopRequireDefault(_xgplayerTransmuxerLoaderFetch);

var _xgplayerUtilsCrypto = require('xgplayer-utils-crypto');

var _xgplayerUtilsCrypto2 = _interopRequireDefault(_xgplayerUtilsCrypto);

var _xgplayerTransmuxerDemuxM3u = require('xgplayer-transmuxer-demux-m3u8');

var _xgplayerTransmuxerDemuxM3u2 = _interopRequireDefault(_xgplayerTransmuxerDemuxM3u);

var _xgplayerTransmuxerDemuxTs = require('xgplayer-transmuxer-demux-ts');

var _xgplayerTransmuxerDemuxTs2 = _interopRequireDefault(_xgplayerTransmuxerDemuxTs);

var _xgplayerTransmuxerBufferPlaylist = require('xgplayer-transmuxer-buffer-playlist');

var _xgplayerTransmuxerBufferPlaylist2 = _interopRequireDefault(_xgplayerTransmuxerBufferPlaylist);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = {
  Tracks: _xgplayerTransmuxerBufferTrack2.default,
  XgBuffer: _xgplayerTransmuxerBufferXgbuffer2.default,
  FetchLoader: _xgplayerTransmuxerLoaderFetch2.default,
  Crypto: _xgplayerUtilsCrypto2.default,
  M3U8Parser: _xgplayerTransmuxerDemuxM3u2.default,
  TsDemuxer: _xgplayerTransmuxerDemuxTs2.default,
  Playlist: _xgplayerTransmuxerBufferPlaylist2.default
};
